[dart] Add dart-server package

duianto opened the pull request:
dart-mode has been split into two packages:
- dart-mode, Provides basic syntax highlighting and indentation.
- dart-server, Factors out the features from dart-mode.el that relied on
external executables, thus improving availability, reliability, testing and
development of dart-mode.el.

Closes #12605

bradyt commented on layers/+lang/dart/packages.el:
```diff
@@ -13,8 +13,11 @@
+    (dart-server :location (recipe :fetcher github :repo bradyt/dart-server))
+    flycheck
+    ))
 
```
Shouldn't this be a one line addition? If I understand correctly, parens on empty lines is not idiomatic.

The trailing parens seem to be explicitly discouraged. See:

> Don't make a habit of putting close-parentheses on lines by themselves; Lisp programmers find this disconcerting.

-- https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html

duianto replied:
I tried it without the surrounding parentheses first (as you suggested in the related issue) but Spacemacs shows these warnings on startup:
```
Warnings:
    - package :location not initialized in layer dart, you may consider removing this
      package from the package list or use the :toggle keyword instead of a `when'
      form.      
    - package recipe not initialized in layer dart, you may consider removing this
      package from the package list or use the :toggle keyword instead of a `when'
      form.      
```

The spaces between packages and keywords seem to work as package delimiters.

The `layers.org` documentation shows that packages that use keywords are surrounded by parentheses.
https://github.com/syl20bnr/spacemacs/blob/develop/doc/LAYERS.org#packagesel

The example also uses trailing parentheses.

That might be why most of the Spacemacs layers used that style.
I unified all layers to use the same style:
Unify packages list parentheses style #11619

But that style seems to be limited to the `<layer>-packages` list, otherwise the Emacs convention seems to be encouraged.

duianto commented:
I updated the commands to use the `server-` infix.

But I don't know how to enable `dart-server` mode when a `.dart` file is opened. `dart-server` mode is disabled with the attempts below.

### Attempt 1
With the latest changes, when `(dart-server)` is in the `dart/init-dart-mode` functions `:init` section, then a warnings buffer opens on startup:

<details><summary>
<code>*Warnings*</code> (Click to expand)
</summary>
<pre>
Error (use-package): gh-md/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): gitignore-templates/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): gnuplot/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): helm/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): hexl/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): image-mode/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): impatient-mode/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): json-navigator/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): markdown-toc/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): org-agenda/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): org-mime/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
Error (use-package): org-pomodoro/:init: Wrong type argument: keymapp, #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"]
</pre>
</details>

### Attempt 2
When `(dart-server)` is moved to a new `:config` section in the `dart/init-dart-mode` function, then there's no warning on startup.

If it's enabled `M-x dart-server RET`, then which-key breaks, when `,` or `SPC` is pressed, then the minibuffer shows:
```
Error running timer ‘which-key--update’: (wrong-type-argument keymapp #[0 "\3062\243\0\307=\203\207\0	\203\207\0\310	\234\311\312\"\311\313\"\311\314\"\204(\0\315\306\316\"\210\204<\0\317c\210\212\320c\210\321 )\315\306\316\"\210\212\317\320\261\210\321 )\310G\316W\203y\0\322`\323\234#\322T\234\323#\324!\324!D\266\202B\262\325\\\262\202K\0\211\237\266\203\326 \266\203\262\202\242\0\327`#\205\242\0
\205\242\0
T\211
GY\203\240\0\310\326 0\207" [last-command dart-server--last-expand-results dart-server--last-expand-parameters-index dart-server--last-expand-index dart-server--last-expand-end dart-server--last-expand-parameters-ranges --cl-block-dart-server-expand-parameters-- dart-server-expand 0 dart-server--get parameterNames defaultArgumentListString defaultArgumentListTextRanges throw nil 40 41 point-marker + 1 copy-marker 2 dart-server--expand-select-parameter < dart-server--last-expand-beginning] 13 ("c:/Users/username/.emacs.d/elpa/26.2/develop/dart-server-20190811.136/dart-server.elc" . 44216) "*"])
```

Disabling `dart-server` mode, `S-M-; (dart-server -1) RET`, restores the which-key functionality.

### Attempt 3
When there's no call to `(dart-server)` in the `dart/init-dart-mode` function, then the same behavior as in `Attempt 2` is observed.

bradyt commented:
I've pushed changes, hopefully this helps?

``` diff
modified   dart-server.el
@@ -304,7 +304,8 @@ Returns nil if `dart-server-sdk-path' is nil."
     (define-key map (kbd "C-c C-t") 'dart-server-find-top-level-decls)
     (define-key map (kbd "C-c C-o") 'dart-server-format)
     (define-key map (kbd "M-/") 'dart-server-expand)
-    (define-key map (kbd "M-?") 'dart-server-expand-parameters))
+    (define-key map (kbd "M-?") 'dart-server-expand-parameters)
+    map)
   "Keymap used in dart-server buffers.")
 
 
@@ -1262,7 +1263,8 @@ Also removes this function from `post-command-hook'."
     (define-key map (kbd "<mouse-2>") nil)
     (define-key map (kbd "C-c C-b") nil)
     (define-key map (kbd "C-c C-c") nil)
-    (define-key map (kbd "C-c C-f") nil))
+    (define-key map (kbd "C-c C-f") nil)
+    map)
   "Keymap used in Dart popup buffers.")
 
 (defun dart-server-do-it-again ()
```

duianto commented:
@bradyt Congratulations on getting the `dart-server` package included in Melpa. 🎉 

I've now updated the PR to pull from melpa.

The latest change fixed the startup warnings when `(dart-server)` is in the `dart/init-dart-mode` functions `:init` section. And there are no which-key errors when `dart-server` is enabled manually. 👍 

I added `(dart-server)` to a use-package `:config` section in `dart/init-dart-server`:
```elisp
(defun dart/init-dart-server ()
  (use-package dart-server
    :config (dart-server)))
```

It works the same as when `(dart-server)` is called from the `dart/init-dart-mode` functions use-package `:init` or `:config` sections, but it seemed to fit best in it's own init-package-name section.

In either case `(dart-server)` is needed for the key binding functions to be loaded when a `.dart` file is opened.

The `dart-server` variable still has the value `nil`, but I might have incorrectly assumed that it needed to be `t`.

When a dart layer key binding is pressed, for example `SPC m ?` is pressed, which calls `dart-server-show-hover`, then the minibuffer shows:
>Starting Dart analysis server.

And the `SPC a p` Process List buffer has a new dart-server entry:
>dart-server-... 13616   run     *dart-server... --           C:/tools/dart-sdk/bin/dart.exe C:/tools/dart-sdk/bin/snapshots/analysis_server.dart.snapshot

So it might be working as expected. Although nothing happens when the dart layer key bindings are pressed (except for `SPC m =` which calls `dart-server-format`, it always works as expected).

Maybe a popup should appear describing the word/symbol under the cursor?
As described in this comment: https://github.com/syl20bnr/spacemacs/issues/11631#issuecomment-483756419

I also noticed that the Spacemacs dart layer readme needed to be updated.
https://github.com/syl20bnr/spacemacs/tree/develop/layers/%2Blang/dart

The variables now use the `server-` infix, and the variable documentation link now points to the dart-server repo.

smile13241324 commented:
This sounds great. Thanks for helping out @bradyt. @duianto should I have another go at testing this?

duianto commented:
@smile13241324 It's probably best since `(dart-server)` is called from the `dart/init-dart-server` function now and I'm not sure how it's supposed to work.

smile13241324 commented:
@duianto @bradyt I have just checked this PR and it seems to work only partially. What I have done.
* Opened sample file test.dart with input:
```dart
int fib(int n) => (n > 2) ? (fib(n - 1) + fib(n - 2)) : 1;
// this is a fibonacci function implementation with a ternary operator in Dart
// this code shall be read as:
// If int n > 2, return fib(n - 1) + fib(n - 2);
// otherwise, return int 1 as result

void main() {
  print('fib(20) = ${fib(20)}');  
}
```
* There I have checked:
   * That all packages for dart-mode are active per default -> works
   * That calling the external services work -> only partially
       * `SPC m =` => formatting, works as expected.
       * `SPC m ?` or any other `SPC m` command fails with 
```org
Dart analysis server crashed, restarting.
error in process filter: json-read: JSON readtable error: 69
error in process filter: JSON readtable error: 69
```   
* Trying to use auto-completion yields the following error message when tried to be used with empty selection:
```org
user-error: No dynamic expansion for ‘20)}');
  ’ found
```
I think the package is correctly integrated into the dart-layer now, but the errors make me suspicious if we have missed something in the setup.

bradyt commented:
@smile13241324 Remember, I don't use spacemacs, so you're pinging me on a reproduce that I can't help with. Please reproduce without spacemacs, and a minimal config, and minimal steps to reproduce. Also, you didn't say which OS you're on, which is unfortunately very relevant to the issues historically filed against this codebase (dart-server.el now, but originally part of dart-mode.el).

bradyt commented:
I don't doubt that dart-server.el is broken. That's part of the motivation for splitting it, not to mention the dependencies needed for even writing tests. But then I would prefer spacemacs users to familiarize themselves with emacs enough to describe issues minimally. Not just for my benefit, but in case someone else wants to help fix dart-server.el.

bradyt commented:
@smile13241324 Can you check if this issue occurs with stable dart-mode, before dart-server was split off?

smile13241324 commented:
Here is my system information, I am on manjaro linux.
Dart-sdk is version 2.4.1.

Yeah I understand, to make sure this is not coming from us I will try to reproduce with vanilla emacs first. I am working so much with Spacemacs I sometimes forget not everybody is using it :) sorry for my premature post @bradyt.

Sure I will also have a try with stable and post the results.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 26.2
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 5e2411c4f)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(lsp dap
     (dart :variables dart-server-sdk-path "~/Downloads/dart-sdk/")
     (helm :variables helm-enable-auto-resize t)
     nim spacemacs-purpose ocaml
     (yang :variables yang-pyang-rules "ietf")
     ietf scheme
     (auto-completion :variables auto-completion-return-key-behavior 'complete auto-completion-tab-key-behavior 'cycle auto-completion-complete-with-key-sequence nil auto-completion-complete-with-key-sequence-delay 0.1 auto-completion-idle-delay 0 auto-completion-private-snippets-directory nil auto-completion-enable-snippets-in-popup t auto-completion-enable-help-tooltip t auto-completion-enable-sort-by-usage t)
     better-defaults
     (clojure :variables clojure-enable-fancify-symbols t clojure-enable-sayid t clojure-enable-clj-refactor t clojure-enable-linters t)
     gpu graphviz
     (plantuml :variables plantuml-jar-path "~/.plantuml/plantUml.jar" org-plantuml-jar-path "~/.plantuml/plantUml.jar")
     salt ibuffer emacs-lisp pdf
     (org :variables org-want-todo-bindings t org-enable-github-support t org-enable-bootstrap-support t org-enable-reveal-js-support t org-enable-hugo-support t org-enable-trello-support t org-enable-epub-support t org-enable-sticky-header nil)
     (shell :variables shell-enable-smart-eshell t shell-default-shell 'ansi-term shell-default-height 30 shell-default-position 'bottom shell-default-full-span nil shell-default-term-shell "/usr/bin/fish")
     syntax-checking version-control import-js web-beautify semantic templates epub themes-megapack
     (c-c++ :variables c-c++-default-mode-for-headers 'c++-mode c-c++-enable-clang-support t c-c++-enable-clang-format-on-save t c-c++-enable-auto-newline t)
     (cmake :variables cmake-enable-cmake-ide-support t)
     git github sml bm
     (copy-as-format :variables copy-as-format-default "markdown" copy-as-format-asciidoc-include-file-name t)
     pass common-lisp
     (python :variables python-backend 'lsp python-test-runner 'pytest python-formatter 'yapf python-format-on-save t python-save-before-test t python-sort-imports-on-save t)
     windows-scripts agda jr shell-scripts prodigy emoji
     (markdown :variables markdown-live-preview-engine 'vmd)
     bibtex
     (json :variables json-fmt-tool 'web-beautify)
     debug
     (latex :variables latex-enable-auto-fill t latex-enable-magic nil latex-enable-folding t)
     lua html
     (javascript :variables javascript-import-tool 'import-js javascript-backend 'lsp javascript-fmt-tool 'web-beautify javascript-repl 'nodejs js2-basic-offset 2 js-indent-level 2)
     react yaml prettier
     (multiple-cursors :variables multiple-cursors-backend 'evil-mc)
     nginx docker csv search-engine
     (restclient :variables restclient-use-org t)
     restructuredtext selectric octave purescript speed-reading systemd imenu-list dap
     (java :variables java-backend 'lsp)
     groovy
     (go :variables go-backend 'lsp go-linter 'golangci-lint gofmt-command "goimports" go-tab-width 4 go-format-before-save t go-use-gocheck-for-testing t go-use-test-args "-race -timeout 10s" godoc-at-point-function 'godoc-gogetdoc)
     coq django
     (elm :variables elm-sort-imports-on-save t elm-format-on-save t)
     idris
     (haskell :variables haskell-completion-backend 'lsp haskell-enable-hindent t haskell-process-type 'stack-ghci)
     pandoc sphinx parinfer asciidoc
     (treemacs :variables treemacs-use-follow-mode t treemacs-use-filewatch-mode t treemacs-use-collapsed-directories 3)
     ansible puppet rebox rust hy xkcd typography vimscript
     (terraform :variables terraform-auto-format-on-save t)
     (ruby :variables ruby-enable-enh-ruby-mode t ruby-backend 'lsp)
     ruby-on-rails asm
     (sql :variables sql-capitalize-keywords t)
     perl6 autohotkey
     (elixir :variables flycheck-elixir-credo-strict t)
     faust vagrant erlang
     (dash :variables helm-dash-docset-newpath "~/.local/share/Zeal/Zeal/docsets")
     games php nginx racket
     (colors :variables colors-colorize-identifiers 'all)
     d
     (typescript :variables typescript-backend 'lsp typescript-fmt-tool 'tide typescript-linter 'tslint typescript-fmt-on-save t))
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG IMAGEMAGICK SOUND GPM DBUS GSETTINGS GLIB NOTIFY ACL GNUTLS LIBXML2 FREETYPE M17N_FLT LIBOTF XFT ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD LCMS2

smile13241324 commented:
I have tried to work with version 1.0.5 from melpa stable. The result is the same on my side, format works but the analysis server crashes with `Process dart-analysis-server exited abnormally with code 254`.

However I assume this comes from my environment as at @duianto's system the analysis server seems to be working.

duianto commented:
>@duianto does the second comment negate your first comment? Or should we expect a minimal reproduce?

I'm not sure which first comment your referring to. But it probably doesn't matter, the issues were related to the configuration in Spacemacs.

@smile13241324 I think I've got it working now, by moving the key bindings and configurations to the `dart/init-dart-server` functions `:init` section, and calling `(dart-mode)` from the `:config` section.

But flycheck doesn't work.
Pressing `SPC e s` which calls `flycheck-select-checker`, and choosing `dart-server-analysis-server`.
Shows:
```
Syntax checker in buffer *Dart Search* in dart-server-popup-mode:

  dart-server-analysis-server
    - major mode: `dart-server-popup-mode' not supported
    - may enable: yes

Flycheck cannot use this syntax checker for this buffer.

Flycheck Mode is disabled. Use SPC u M-x flycheck-disable-checker to enable
disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20190821.1725)
Emacs version:    26.2
System:           x86_64-w64-mingw32
Window system:    w32
```

The suggestion, at least for Windows might be to use lsp:
https://github.com/bradyt/dart-server/issues/4#issuecomment-447178843

---

I might also have found a possible reason why the server crashes for you.
It doesn't crash for me, in either Ubuntu or Windows, but the following happens both with or without Spacemacs.

Don't do this if you have unsaved files open, because Emacs becomes slow, and if one waits long enough then it's not impossible that it crashes, like it does for you.
And kill the `dart-server-analysis-server` buffer after noticing that it keeps increasing in size.

If `dart-server-find-refs` (`,ff`) is called on the instances of `int` in this example `fibonacci.dart` file:
```dart
// https://dart.dev/samples#functions
int fibonacci(int n) {
  if (n == 0 || n == 1) return n;
  return fibonacci(n - 1) + fibonacci(n - 2);
}

var result = fibonacci(20);
```
then three buffers open, the first one opens in another window, the other two open in the background:
- `*Dart Search*`
- `*dart-server-analysis-server*`
- `dart-server-analysis-server`

`*Dart Search*` contains: `References to int:`
`*dart-server-analysis-server*` is empty.
But `dart-server-analysis-server` keeps growing in size.

This can be seen in Spacemacs by opening HELM Mini (`SPC b b`) and adding a space, then erasing it, to update the buffer list. 

The `dart-server-analysis-server` buffer keeps filling with data like this:
```
{"event":"search.results","params":{"id":"0","results":[{"location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":0,"length":3,"startLine":1,"startColumn":1},"kind":"REFERENCE","isPotential":false,"path":[{"kind":"FUNCTION","name":"fibonacci","location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":4,"length":9,"startLine":1,"startColumn":5},"flags":8,"parameters":"(int n)","returnType":"int"},{"kind":"COMPILATION_UNIT","name":"fibonacci.dart","location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":0,"length":0,"startLine":1,"startColumn":1},"flags":16},{"kind":"LIBRARY","name":"","location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":0,"length":0,"startLine":1,"startColumn":1},"flags":0}]},{"location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":14,"length":3,"startLine":1,"startColumn":15},"kind":"REFERENCE","isPotential":false,"path":[{"kind":"PARAMETER","name":"n","location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":18,"length":1,"startLine":1,"startColumn":19},"flags":0,"returnType":"int"},{"kind":"FUNCTION","name":"fibonacci","location":{"file":"c:\\Users\\username\\Temp\\fibonacci.dart","offset":4,"length":9,"startLine":1,"startColumn":5},"flags":8,"parameters":"(int n)","returnType":"int"}
```

While testing I might have waited about 10-15 seconds before being able to select the whole buffer and coping it's current contents. In that time the buffer had filled to over two million characters.

bradyt commented:
I see issues with even the dart formatter burning CPU, at https://github.com/bradyt/dart-server/issues/10. I'd recommend not adding dart-server.el to the layer at all, and instead using reformatter.el and eglot or lsp-mode instead.

smile13241324 commented:
@duianto I have just merged the first version of the LSP integration for dap from here #12753. I think we should try to stick with the LSP solution rather than trying to get the standalone dart-server thing to run properly.

Would you mind having a look at the current state in the dart-layer and adjust this PR if necessary? I am not sure if the new dart-server package is actually still needed, but I am far more optimistic with the LSP server then what I am with the dart-server package right now.

duianto commented:
#### Without this PR:
Only the new flutter key binding `SPC m x x` works.

None of the other major mode key bindings under `SPC m` work, because they are still referring to the commands from before the split: `dart-mode` / `dart-server` .

The same thing applies to the variables that are listed in the `dart` layers readme:
https://github.com/syl20bnr/spacemacs/tree/develop/layers/%2Blang/dart

#### With this PR
With my limited experience with dart (from making/testing this PR), all of the `SPC m` key bindings seems to work as expected.

The issue with with pressing `SPC m f f` on an instance of `int`, as I mentioned in the previous comment, still occurs.

I've reported it upstream:
dart-server-find-refs on int opens a dart buffer that keeps growing in size https://github.com/bradyt/dart-server/issues/11

I suspect that Emacs eventually will slow to a crawl or even crash if one waits long enough and the `dart-server-analysis-server` buffer keeps filling with characters.

#### A possible solution
I disabled the `SPC m f f` key binding with a comment linking to the upstream issue.

#### Notes
I included a `flutter` which-key prefix and fixed a trailing parenthese as well.

#### Conclusion
Now this PR seems to bring the `dart` layer back into a working state.

smile13241324 commented:
Thank you for contributing to spacemacs :purple_heart:, I have cherry picked your commit into develop.
